### ISSUE TITLE
Is there a less painful way to implement deserialize for structs

### ISSUE THREAD

**Issue by elliottslaughter** [serde-rs/serde#770]
I'm currently staring at https://serde.rs/deserialize-struct.html trying to figure out if there is a less painful way to get where I want.

Here's my problem: I'm trying to implement JSON serialization for [Pandoc ASTs](https://hackage.haskell.org/package/pandoc-types-1.17.0.5/docs/Text-Pandoc-Definition.html). The datatypes themselves ported over easily. And I can serialize them easily, as long as I don't care about the format. But getting them to serialize the way Pandoc serializes them (which is the entire point of this endeavor) is turning into a pain.

Here's an example of what I mean. This is an except of the types I want to serialize:

```
pub enum Block {
    Para(Vec<Inline>),
    ...
}
pub enum Inline {
    Str(String),
    ...
}
```

And here is the representation Pandoc uses:

```
{"t":"Para","c":[{"t":"Str","c":"b"}]}
```

Whereas here is the default representation that serde_json chooses:

```
{"Para":[{"Str":"b"}]}
```

In other words, serde_json chooses to encode enums as `{"Constructor": [args]}` whereas Pandoc encodes this as `{"t": "Constructor": "c": [args]}`.

I briefly played around with the idea of serializing types by maintaining an extra type `SerializedX` for each type `X` and manipulating the format of `SerializedX` to match what Pandoc wants to see. I managed to make this [kind of sort of work for some types](https://github.com/elliottslaughter/rust-pandoc-types/blob/675aed7841ad216af394b13f26437be1cd5df738/src/definition.rs#L35-L75), although at the expense of copying everything that gets serialized/deserialized.

However, I can't make it work at for the case above, because the constructor name determines what I need to serialize as the args for each branch of the enum. The code, such as it is, [can be found here](https://github.com/elliottslaughter/rust-pandoc-types/blob/675aed7841ad216af394b13f26437be1cd5df738/src/definition.rs#L108-L153).

I was hoping to avoid the manual deserialization approach, because that seems like a lot of code to write for something that shouldn't be that complicated. In fact, the format Pandoc likes to use is really only a minor variation on what serde_json already does. It seems like I could write a code generator do what I want, though I have no idea if this would be more or less painful than just writing out all the code.

Are there any other options? Sorry if this seems like a vague question; I guess I'm mainly just trying to assess the landscape and see if there are any approaches I'm not aware of.

Thanks.

**Comment by dtolnay** [comment]
Have you considered using an [internally tagged enum](https://serde.rs/enum-representations.html#internally-tagged)?

It is a bit repetitive because it makes you repeat the key "c" for all of the variants but it definitely beats handwriting a Deserialize impl. We are planning to add better support for this Haskell-style representation. It is tracked in #758 and will look something like `#[serde(tag = "t", content = "c")]`. Would love a PR if you have the time and energy!

```rust
#[macro_use]
extern crate serde_derive;
extern crate serde_json;

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "t")]
pub enum Block {
    Para { c: Vec<Inline> },
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "t")]
pub enum Inline {
    Str { c: String },
}

fn main() {
    let blk = Block::Para { c: vec![Inline::Str { c: "b".to_owned() }] };

    let j = serde_json::to_string(&blk).unwrap();
    println!("{}", j);

    let de: Block = serde_json::from_str(&j).unwrap();
    println!("{:#?}", de);
}
```

**Comment by dtolnay** [comment]
The Pandoc representation has been implemented in #771 and #782.

```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "t", content = "c")]
pub enum Block {
    Para(Vec<Inline>),
}

#[derive(Serialize, Deserialize)]
#[serde(tag = "t", content = "c")]
pub enum Inline {
    Str(String),
}
```
